nilearn/decoding/tests/test_decoder.py

```diff
@@ -1089,3 +1079,156 @@ def test_decoder_decision_function_raises_value_error(
+    """Test _apply_mask on surface image."""
+    X, _ = _make_surface_class_data()
+    model = Decoder(mask=SurfaceMasker())
+    X_masked = model._apply_mask(X)
+
+    assert X_masked.shape == X.shape
+    assert type(model.mask_img_).__name__ == "SurfaceImage"
+
+
+def test_decoder_screening_percentile_surface_default(
+    _make_surface_class_data,
+):
+    """Test default screening percentile with surface image."""
+    warnings.simplefilter("ignore", ConvergenceWarning)
+    X, y = _make_surface_class_data()
+
+    model = Decoder(mask=SurfaceMasker())
+    model.fit(X, y)
+    assert model.screening_percentile_ == 20
+
+
+@pytest.mark.parametrize("perc", [None, 100, 0])
+def test_decoder_screening_percentile_surface(perc, _make_surface_class_data):
+    """Test passing screening percentile with surface image."""
+    warnings.simplefilter("ignore", ConvergenceWarning)
+    X, y = _make_surface_class_data()
+
+    model = Decoder(mask=SurfaceMasker(), screening_percentile=perc)
+    model.fit(X, y)
+    if perc is None:
+        assert model.screening_percentile_ == 100
+    else:
+        assert model.screening_percentile_ == perc
+
+
+@pytest.mark.parametrize("mask", [None, SurfaceMasker()])
+@pytest.mark.parametrize("decoder", [_BaseDecoder, Decoder, DecoderRegressor])
+def test_decoder_fit_surface(decoder, _make_surface_class_data, mask):
+    """Test fit for surface image."""
+    warnings.simplefilter("ignore", ConvergenceWarning)
+    X, y = _make_surface_class_data()
+    model = decoder(mask=mask)
+    model.fit(X, y)
+
+    assert model.coef_ is not None
+
+
+@pytest.mark.parametrize("decoder", [_BaseDecoder, Decoder, DecoderRegressor])
+def test_decoder_fit_surface_with_mask_image(
+    _make_surface_class_data, decoder, mini_mask
+):
+    """Test fit for surface image."""
+    warnings.simplefilter("ignore", ConvergenceWarning)
+    X, y = _make_surface_class_data()
+    model = decoder(mask=mini_mask)
+    model.fit(X, y)
+
+    assert model.coef_ is not None
+
+
+@pytest.mark.parametrize("decoder", [_BaseDecoder, Decoder, DecoderRegressor])
+def test_decoder_error_incompatible_surface_mask_and_volume_data(
+    decoder, mini_mask, tiny_binary_classification_data
+):
+    """Test error when fitting volume data with a surface mask."""
+    data_volume, y, _ = tiny_binary_classification_data
+    model = decoder(mask=mini_mask)
+    model.fit(data_volume, y)
+
+    model = decoder(mask=SurfaceMasker())
+    model.fit(data_volume, y)
+
+
+@pytest.mark.parametrize("decoder", [_BaseDecoder, Decoder, DecoderRegressor])
+def test_decoder_error_incompatible_surface_data_and_volume_mask(
+    _make_surface_class_data, decoder, tiny_binary_classification_data
+):
+    """Test error when fiting for surface data with a volume mask."""
+    data_surface, y = _make_surface_class_data()
+    _, _, mask = tiny_binary_classification_data
+    model = decoder(mask=mask)
+    model.fit(data_surface, y)
```

Comment:
- Those are currently failing and I would prefer to merge https://github.com/nilearn/nilearn/pull/4603 first so its content can be reused to validate the compatibility between data and maskers.